### PR TITLE
ExistingAutocompleter should check for canEdit() permission

### PR DESCRIPTION
**Pull request by uniun**
GridFieldAddExistingAutocompleter should check for canEdit() permission.

### CODE REVIEW

**Comment by dhensby** [comment]
There's now a test failure, please can you look into it and provide a test for this? thanks

**Comment by dhensby** [comment]
@uniun are you likely to make any progress on the PR?

Also, should the canEdit permission dictate if you can add the item as a relation? It's not really an "edit" as such just an association.

**Comment by uniun** [comment]
Sorry, I've been quite busy lately. I'll look on it.
And yes, I agree this is not an "edit". But it should check for canView at least I think.

**Comment by dhensby** [comment]
No problem

Definitely canView... but typically that'll be true because it should be used to affect if it is shown on the front end of the site :/

Really tricky as actually there's probably a new permission required for this kind of thing. Maybe a config option needed to say whether to check a permission at all and, if so, which one?

**Comment by uniun** [comment]
Maybe. But how about GridFieldViewButton then? This component also checks for canView.

I understand typically that'll true and it should be used on frontend of the site. But sometimes this permission is usefull on backend too.
For example, if you want to "hide" members of "Administrator" group (or any other) from other members it can be achieved with canView.

**Comment by dhensby** [comment]
> Maybe. But how about GridFieldViewButton then? This component also checks for canView.

Ok, lets use `canView` then

> For example, if you want to "hide" members of "Administrator" group (or any other) from other members it can be achieved with canView.

good point and this is an example where it would be very useful

**Comment by uniun** [comment]
Sorry, but the test still fails. This is the first time I write/edit the test :)

**Comment by uniun** [comment]
So what what should I do now?
1. add canView() to the GridFieldTest_Team object in GridFieldTest.php
2. change team object in GridFieldAddExistingAutocompleterTest to a new object, e.g. GridFieldAddExistingAutocompleterTest _Team?

File: tests/forms/gridfield/GridFieldAddExistingAutocompleterTest.php [hunk 1]
```diff
@@ -111,3 +111,16 @@ public function Form() {
 		return new Form($this, 'Form', new FieldList($field), new FieldList());
 	}
 }
+
+class GridFieldTest_Team extends DataObject implements TestOnly {
```
Comment (dhensby) [inline tests/forms/gridfield/GridFieldAddExistingAutocompleterTest.php]: This class already exists, it seems. Either you need to pick a new name or you should just amend the existing class to have the canView function you need (assuming that doesn't break other tests)!

**Comment by uniun** [comment]
I've decided to create new DataObject for this test. I don't want to change default team object because it may brake other test. Also any other test (e.g. GridFieldExportButtonTest) that checks canView permission, do not use default team object.

**Comment by uniun** [comment]
Damn, test still fails... But now there is something GridFieldPaginatorTest? I didn't change it.

**Comment by dhensby** [comment]
That test still needs the `GridFieldTest_Cheerleader` object and that needs to be defined in the `$extraDataObjects` array for the `GridFieldPaginatorTest` test

**Comment by dhensby** [comment]
Ah so, you've renamed classes that were relied upon in another test. Those classes need to remain really, you only need to create one new class for this test rather than rename all of the existing ones.

**Comment by uniun** [comment]
Hmm. I think I need to create at least two of them: GridFieldTest_Team and GridFieldTest_Player. 
$field = new GridField('testfield', 'testfield', $player->Teams(), $config);
Gridfield uses player's teams, so I need to change Teams relation model to new one.

**Pull request closed without merging**